packages/noya-state/src/selectors/gradientSelectors.ts

```diff
@@ -0,0 +1,203 @@
+    (styleType === 'border' &&
+      layer.style?.borders?.[fillIndex].fillType !== Sketch.FillType.Gradient)
+  )
+    return;
+
+  const transform = getLayerTransformAtIndexPath(
+    page,
+    indexPath,
+    AffineTransform.identity,
+    'includeLast',
+  ).scale(layer.frame.width, layer.frame.height);
+
+  const gradient =
+    styleType === 'fill'
+      ? layer.style?.fills?.[fillIndex].gradient
+      : layer.style?.borders?.[fillIndex].gradient;
+
+  if (!gradient) return;
+
+  const from = PointString.decode(gradient.from);
+  const to = PointString.decode(gradient.to);
+
+  const extremePoints = {
+    from: transform.applyTo(from),
+    to: transform.applyTo(to),
+  };
+
+  const stops = sorted
+    ? [...gradient.stops].sort((a, b) => a.position - b.position)
+    : gradient.stops;
+
+  return stops.map((stop) => {
+    return {
+      color: stop.color,
+      point: {
+        x: lerp(extremePoints.from.x, extremePoints.to.x, stop.position),
+        y: lerp(extremePoints.from.y, extremePoints.to.y, stop.position),
+      },
+    };
+  });
+}
+
+/**
+ * This function normalizes a gradient so that gradient stop positions are
+ * always between 0 and 1. The first position will be 0 and the last will be 1.
+ * This function shouldn't change the gradient visually at all.
+ *
+ * We need to call this when we finish editing a gradient, or before dragging
+ * a gradient stop on the canvas. It's OK if a gradient is temporarily in a state
+ * where the stop positions aren't correct, since this is a better experience when
+ * editing with the gradient picker.
+ */
+export function fixGradientPositions(gradient: Sketch.Gradient) {
+  const stops = gradient.stops;
+
+  const sorted = [...stops].sort((a, b) => a.position - b.position);
+
+  const from = PointString.decode(gradient.from);
+  const to = PointString.decode(gradient.to);
+
+  const stopPoints = sorted.map(
+    (stop): GradientStopPoint => ({
+      point: {
+        x: lerp(from.x, to.x, stop.position),
+        y: lerp(from.y, to.y, stop.position),
+      },
+      color: stop.color,
+    }),
+  );
+
+  const newFrom = stopPoints[0].point;
+  const newTo = stopPoints[stopPoints.length - 1].point;
+
+  gradient.from = PointString.encode(newFrom);
+  gradient.to = PointString.encode(newTo);
+  gradient.stops = stopPoints.map((stopPoint) => {
+    return {
+      _class: Sketch.ClassValue.GradientStop,
+      color: stopPoint.color,
+      position: getLinePercentage(stopPoint.point, [newFrom, newTo]),
+    };
+  });
+}
+
+export function getGradientStopIndexAtPoint(
+  state: ApplicationState,
+  point: Point,
+): number {
+  const selectedLayerGradientPoints = getSelectedGradientStopPoints(
+    state,
+    true,
+  );
+
+  if (!selectedLayerGradientPoints) return -1;
+
+  return selectedLayerGradientPoints.findIndex((gradientPoint) =>
+    isPointInRange(gradientPoint.point, point, 5.5),
+  );
+}
+
+function isPointOnLine(A: Point, B: Point, point: Point) {
```

Comment:
Let's move this into `line.ts` in `noya-geometry` and add a couple tests for it.

Reply:
Is this equivalent? If so I think it's a little clearer

```ts
return Math.abs(d1 + d2 - lineLength) < 0.5
```

Reply:
We should change the argument types to `function isPointOnLine(point: Point, line: [Point, Point])` to be a little clearer and consistent with `getClosestPointOnLine`. And rename `lineLen` => `lineLength`, since we don't usually use abbreviations